### ISSUE TITLE
mingw-w64-x86_64-winpthreads-git-8.0.0.5904.1278c953-1 is broken

### ISSUE THREAD

**Issue by zufuliu** [msys2/MINGW-packages#6530]
I get following errors when building https://github.com/zufuliu/notepad2

```
D:/Dev/gcc/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/Dev/gcc/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/lib/../lib\libmsvcrt.a(lib64_libmsvcrt_os_a-mbrtowc.o): in function `__mbrtowc_cp':
D:/mingwbuild/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/misc/mbrtowc.c:60: undefined reference to `__imp_IsDBCSLeadByteEx'
D:/Dev/gcc/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/mingwbuild/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/misc/mbrtowc.c:50: undefined reference to `__imp_MultiByteToWideChar'
D:/Dev/gcc/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/mingwbuild/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/misc/mbrtowc.c:85: undefined reference to `__imp_MultiByteToWideChar'
D:/Dev/gcc/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/Dev/gcc/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/lib/../lib\libmsvcrt.a(lib64_libmsvcrt_os_a-wcrtomb.o): in function `__wcrtomb_cp':
D:/mingwbuild/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/misc/wcrtomb.c:35: undefined reference to `__imp_WideCharToMultiByte'
```

**Comment by Alexpux** [comment]
@zufuliu report it to mingw-w64 mailing list

**Comment by Biswa96** [comment]
cc @mstorsjo

**Comment by zufuliu** [comment]
Added a ticket at https://sourceforge.net/p/mingw-w64/bugs/842/

**Comment by zufuliu** [comment]
the broken package is actually mingw-w64-x86_64-winpthreads-git-8.0.0.5904.1278c953-1
revert back to mingw-w64-x86_64-winpthreads-git-8.0.0.5814.9dbf4cc1-1
link with `-static` works, so the new winpthreads (5904) can't linked with `-static` ?

**Comment by lhmouse** [comment]
Please add `-D__USE_MINGW_ANSI_STDIO=0` into `CPPFLAGS` when building **winpthreads**.

**winpthreads** contains calls to `printf()` etc, which used to be imported from **MSVCRT**. Due to recent changes, the broken library calls the **mingw-w64** one. This behavior should be suppressed when building it, as the CRT is not available.

**Comment by zufuliu** [comment]
@lhmouse  see https://sourceforge.net/p/mingw-w64/bugs/842/#0106
it's pure win32 codes with a few crt functions, no pthreads.
The code are at https://github.com/zufuliu/notepad2/tree/master/metapath/src

**Comment by lhmouse** [comment]
@zufuliu Try adding `-D__USE_MINGW_ANSI_STDIO=0` into `CPPFLAGS`.

**Comment by zufuliu** [comment]
@lhmouse it not help.

```batch
@echo off
SET "DEFINES=-D_WINDOWS -DUNICODE -D_UNICODE -DNDEBUG -D_WIN64 -D_WIN32_WINNT=0x0502 -DWINVER=0x0502"
SET "FLAGS=-municode -Wall -Wextra -Wshadow -Wimplicit-fallthrough -Wformat=2 -O2 -ffast-math -fomit-frame-pointer"
SET "CFLAGS=-std=gnu11 %FLAGS% %DEFINES% -D__USE_MINGW_ANSI_STDIO=0"
SET "CXXFLAGS=-std=gnu++17 -fno-rtti %FLAGS% %DEFINES% -D__USE_MINGW_ANSI_STDIO=0"
SET "LDLIBS=-lkernel32 -luser32 -lgdi32 -lcomctl32 -lcomdlg32 -ladvapi32 -lshlwapi -lshell32 -lole32 -loleaut32 -luuid -luxtheme -lpsapi"
SET "LDFLAGS=-mwindows -municode -s -O2 -static -Wl,--gc-sections -ffast-math -fno-rtti -fomit-frame-pointer"

SET "OBJS=Dialogs.o Dlapi.o DropSource.o Helpers.o metapath.o metapath.rc.o"
DEL /Q *.o

gcc.exe -c Dialogs.c %CFLAGS% -o Dialogs.o
gcc.exe -c Dlapi.c %CFLAGS% -o Dlapi.o
g++.exe -c DropSource.cpp %CXXFLAGS% -o DropSource.o
gcc.exe -c Helpers.c %CFLAGS% -o Helpers.o
gcc.exe -c metapath.c %CFLAGS% -o metapath.o
windres.exe metapath.rc -DRC_INVOKED -DUNICODE -D_UNICODE  -D_WIN64 -o metapath.rc.o

g++.exe -o metapath.exe %OBJS% %LDLIBS% %LDFLAGS%
pause
```

**Comment by lhmouse** [comment]
Probably it's not the solution. There is a missing `-lkernel32` after `-lmsvcrt` in GCC's builtin specs, which should be patched.

[0001-gcc-config-i386-mingw32.h-Ensure-lmsvcrt-precede-lke.patch.txt](https://github.com/msys2/MINGW-packages/files/4701944/0001-gcc-config-i386-mingw32.h-Ensure-lmsvcrt-precede-lke.patch.txt)

**Comment by zufuliu** [comment]
manually use it also not work.
```batch
SET "LDLIBS=-lmoldname -lmingwex -lmsvcrt -lkernel32 -luser32 -lgdi32 -lcomctl32 -lcomdlg32 -ladvapi32 -lshlwapi -lshell32 -lole32 -loleaut32 -luuid -luxtheme -lpsapi"
```

you can clone the code and test above with build batch script (run it in metapath/src folder),
`git clone --depth=1 https://github.com/zufuliu/notepad2.git`

**Comment by mati865** [comment]
Running GCC with `-v` shows:
```
 D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/collect2.exe -plugin D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/liblto_plugin-0.dll -plugin-opt=D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/lto-wrapper.exe -plugin-opt=-fresolution=D:\msys64\tmp\ccNj6zCx.res -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_eh -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-lgdi32 -plugin-opt=-pass-through=-lcomdlg32 -plugin-opt=-pass-through=-ladvapi32 -plugin-opt=-pass-through=-lshell32 -plugin-opt=-pass-through=-luser32 -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_eh -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -m i386pep --subsystem windows -Bstatic -o metapath.exe -s D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/lib/../lib/crt2u.o D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/crtbegin.o -LD:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0 -LD:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/lib/../lib -LD:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../lib -LD:/msys64/mingw64/x86_64-w64-mingw32/lib/../lib -LD:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/lib -LD:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/../../.. -LD:/msys64/mingw64/x86_64-w64-mingw32/lib Dialogs.o Dlapi.o DropSource.o Helpers.o metapath.o metapath.rc.o -lmoldname -lmingwex -lmsvcrt -lkernel32 -luser32 -lgdi32 -lcomctl32 -lcomdlg32 -ladvapi32 -lshlwapi -lshell32 -lole32 -loleaut32 -luuid -luxtheme -lpsapi --gc-sections -lstdc++ -lmingw32 -lgcc -lgcc_eh -lmoldname -lmingwex -lmsvcrt -lpthread -lgdi32 -lcomdlg32 -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc -lgcc_eh -lmoldname -lmingwex -lmsvcrt D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/crtfastmath.o D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/lib/../lib/default-manifest.o D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.1.0/crtend.o
```
I suppose after `-lpthread` there should be `-lmsvcrt` followed by `-lkernel32`. Instead there is `-lkernel32` followed by `-lmsvcrt`.

As workaround you can add libs requiring pthread first to `LDLIBS`: `-lstdc++ -lgcc_eh -lpthread`.

With that change it built fine with that script:
```bash
#!/bin/bash
DEFINES="-D_WINDOWS -DUNICODE -D_UNICODE -DNDEBUG -D_WIN64 -D_WIN32_WINNT=0x0502 -DWINVER=0x0502"
FLAGS="-municode -Wall -Wextra -Wshadow -Wimplicit-fallthrough -Wformat=2 -O2 -ffast-math -fomit-frame-pointer"
CFLAGS="-std=gnu11 $FLAGS $DEFINES -D__USE_MINGW_ANSI_STDIO=0"
CXXFLAGS="-std=gnu++17 -fno-rtti $FLAGS $DEFINES -D__USE_MINGW_ANSI_STDIO=0"
LDLIBS="-lstdc++ -lgcc_eh -lpthread -lmoldname -lmingwex -lmsvcrt -lkernel32 -luser32 -lgdi32 -lcomctl32 -lcomdlg32 -ladvapi32 -lshlwapi -lshell32 -lole32 -loleaut32 -luuid -luxtheme -lpsapi"
LDFLAGS="-mwindows -municode -s -O2 -static -Wl,--gc-sections -ffast-math -fno-rtti -fomit-frame-pointer"

OBJS="Dialogs.o Dlapi.o DropSource.o Helpers.o metapath.o metapath.rc.o"
rm -rf *.o

gcc.exe -c Dialogs.c $CFLAGS -o Dialogs.o
gcc.exe -c Dlapi.c $CFLAGS -o Dlapi.o
g++.exe -c DropSource.cpp $CXXFLAGS -o DropSource.o
gcc.exe -c Helpers.c $CFLAGS -o Helpers.o
gcc.exe -c metapath.c $CFLAGS -o metapath.o
windres.exe metapath.rc -DRC_INVOKED -DUNICODE -D_UNICODE  -D_WIN64 -o metapath.rc.o

g++.exe -o metapath.exe $OBJS $LDLIBS $LDFLAGS
```
The same issue does **not** happen with Clang after you remove `-lmsvcrt` from LDLIBS (it doesn't add duplicate `msvcrt` on the command line if it's already present).

**Comment by lhmouse** [comment]
https://www.mail-archive.com/gcc-patches@gcc.gnu.org/msg236889.html

**Comment by zufuliu** [comment]
@mati865 thanks, it works with `-lstdc++ -lgcc_eh -lpthread` added before `-lkernel32`:
`LDLIBS=-lstdc++ -lgcc_eh -lpthread -lkernel32 -luser32 ...`

**Comment by lhmouse** [comment]
As a permanent solution, I have proposed a patch to disable mingw-w64 stdio functions for winpthreads.

**Comment by zufuliu** [comment]
I confirm this was fixed by gcc-10.1.0-3 and winpthreads-git-8.0.0.5906.c9a21571.
